llvm/test/CodeGen/AArch64/vec3-loads-ext-trunc-stores.ll

```diff
@@ -281,23 +279,19 @@ entry:
 define void @store_trunc_add_from_64bits(ptr %src, ptr %dst) {
 ; CHECK-LABEL: store_trunc_add_from_64bits:
 ; CHECK:       ; %bb.0: ; %entry
-; CHECK-NEXT:    sub sp, sp, #16
-; CHECK-NEXT:    .cfi_def_cfa_offset 16
 ; CHECK-NEXT:    ldr s0, [x0]
 ; CHECK-NEXT:    add x9, x0, #4
 ; CHECK-NEXT:  Lloh0:
 ; CHECK-NEXT:    adrp x8, lCPI7_0@PAGE
 ; CHECK-NEXT:  Lloh1:
 ; CHECK-NEXT:    ldr d1, [x8, lCPI7_0@PAGEOFF]
+; CHECK-NEXT:    add x8, x1, #1
 ; CHECK-NEXT:    ld1.h { v0 }[2], [x9]
+; CHECK-NEXT:    add x9, x1, #2
 ; CHECK-NEXT:    add.4h v0, v0, v1
-; CHECK-NEXT:    xtn.8b v1, v0
-; CHECK-NEXT:    umov.h w8, v0[2]
-; CHECK-NEXT:    str s1, [sp, #12]
-; CHECK-NEXT:    ldrh w9, [sp, #12]
```

Comment:
Spent a little time looking at why the default code is so horrible; the primary issue is actually the way the legalizer (GenWidenVectorStores) is trying to lower the operation into an i16 store followed by an i8 store. It ends up generating a bitcast from v4i8 to v2i16, and the default handling for that is completely terrible (it doesn't know how to use a shuffle, so it goes through a stack temporary).

Maybe worth looking into improving the bitcast situation if you're going to continue looking at very narrow vector types.

Reply:
Yep, I am planning to look into this as well.

Reply:
Actually, I am not entirely sure what the exact issue is; the bit cast itself should be a no-op. Perhaps the issue is that the vector types aren't legal? 

It seems like we are going from v3i8 -> v4i8, which isn't legal, and then going from v4i8 -> v4i16 which is legal. Perhaps it would be better to go from v3i8 to v8i8, but I couldn't find a way to do that; setting custom actions require from & to types to by MVTs, but v3i8 isn't a simple value type unfortunately.

Reply:
v4i8 and v2i16 aren't legal types.  And the legalization rule we use for them is "promote", so the bitcast isn't a no-op. There's padding between the elements.  So we need to do a shuffle... but neither AArch64TargetLowering::ReplaceBITCASTResults nor DAGTypeLegalizer::PromoteIntRes_BITCAST has code to do that, so we end up using CreateStackStoreLoad.

The legalization rule for vector types is under the control of the target; see AArch64TargetLowering::getPreferredVectorAction.  You can use that to force v4i8->v8i8.  But changing that impacts a lot of code.

Reply:
Thanks, I'll look into that separately.